File: Sorting/BinarySearch.java
import java.util.Scanner;

public class BinarySearch {

    // Binary search method
    public static int binarySearch(int[] array, int target) {
        int left = 0;
        int right = array.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;  // Find the middle element

            // Check if target is present at mid
            if (array[mid] == target) {
                return mid;
            }
            // If target is greater, ignore left half
            if (array[mid] < target) {
                left = mid + 1;
            }
            // If target is smaller, ignore right half
            else {
                right = mid - 1;
            }
        }
        // Target not found
        return -1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Taking input for the array size
        System.out.print("Enter the number of elements in the array: ");
        int size = scanner.nextInt();

        int[] array = new int[size];

        // Taking input for array elements (assuming sorted array)
        System.out.println("Enter " + size + " sorted integers: ");
        for (int i = 0; i < size; i++) {
            array[i] = scanner.nextInt();
        }

        // Taking input for the target element
        System.out.print("Enter the target element to search: ");
        int target = scanner.nextInt();

        // Perform binary search
        int result = binarySearch(array, target);

        // Output the result
        if (result == -1) {
            System.out.println("Element not present in the array.");
        } else {
            System.out.println("Element found at index " + result);
        }

        scanner.close();
    }
}

File: Sorting/BinarySearch.cpp
#include <iostream>
#include <vector>
#include <algorithm> 

int binarySearch(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2; 

        // Check if the target is present at mid
        if (arr[mid] == target) {
            return mid; 
        }
        // If target is greater, ignore the left half
        else if (arr[mid] < target) {
            left = mid + 1;
        }
        // If target is smaller, ignore the right half
        else {
            right = mid - 1;
        }
    }

    return -1; // Target not found
}

int main() {
    std::vector<int> arr = {34, 12, 56, 78, 23, 7, 89, 1};
    std::sort(arr.begin(), arr.end()); // Sort the array before applying binary search

    int target = 56;
    int result = binarySearch(arr, target);

    if (result != -1) {
        std::cout << "Element found at index " << result << std::endl;
    } else {
        std::cout << "Element not found in the array" << std::endl;
    }

    return 0;
}
